Add tooltip for identicon

dj-pundir opened the pull request:
This is the resolution of issue https://github.com/ChainSafe/Multix/issues/92

Created new class `MultixIdenticon` for tooltip purpose.

I have done following more change:
- Remove theme and size properties : Because most of the places they are using the same values so i made them optional parameters in the control and uses fixed values `ICON_THEME, ICON_SIZE` if not pass.
- Please check the `DEFAULT_AUTO_HIDE_DURATION` time as 2 secs look ok for me as you explained in the requirement it should be around 5 secs. Please change if you want it otherwise.
- Placement can be `bottom` also which can be changed by `DEFAULT_PLACEMENT`



![image](https://github.com/ChainSafe/Multix/assets/61174028/a532a6ed-e999-42ef-9bca-307c98fa5187)

Tbaut commented on packages/ui/src/components/AccountSelect/index.tsx:
```diff
@@ -36,10 +36,9 @@ export const AccountSelect: React.FC<Props> = ({ anchorEl, onClose, ...props })
-                <Identicon
+                <MultixIdenticon
                   value={account.address}
                   theme="substrate"
```
We should remove the theme here too I believe
```suggestion
```

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+}
+const MultixIdenticon = ({ value, theme = ICON_THEME, size = ICON_SIZE, className }: Props) => {
+  const [open, setOpen] = React.useState(false)
+  const handleTooltipClose = () => setOpen(false)
```
Please use `useCallback`here and below
```suggestion
  const handleTooltipClose = useCallback(() => setOpen(false), [])
```

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+  className?: string
+}
+const MultixIdenticon = ({ value, theme = ICON_THEME, size = ICON_SIZE, className }: Props) => {
+  const [open, setOpen] = React.useState(false)
```
Let's import the `useState` from `react` directly.
```suggestion
  const [open, setOpen] = useState(false)
```

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+import React from 'react'
```
```suggestion
import { useState, useCallback } from 'react'
```

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+const MultixIdenticon = ({ value, theme = ICON_THEME, size = ICON_SIZE, className }: Props) => {
+  const [open, setOpen] = React.useState(false)
+  const handleTooltipClose = () => setOpen(false)
+  const handleTooltipOpen = () => setOpen(true)
```
using useCallback here too
```suggestion
  const handleTooltipOpen = useCallback(() => setOpen(true), [])
```

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+      title={DEFAULT_TITLE}
+      placement={DEFAULT_PLACEMENT}
+      leaveDelay={DEFAULT_AUTO_HIDE_DURATION}
+      arrow
```
I'd suggest to remove the arrow.
```suggestion
```

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+      leaveDelay={DEFAULT_AUTO_HIDE_DURATION}
+      arrow
+    >
+      <IconButton onClick={handleTooltipOpen}>
```
Is this needed or can we have the `onClick` handler on the Identicon directly?

Tbaut replied:
I'd appreciate an answer to this question

dj-pundir replied:
Hey @Tbaut  sorry i guess i missed this.
So if we are using tooltip on `Icon` then we have to create a button of that `Icon` and need a wrapper around (using `<IconButton>` ) to use `Tooltip` but for Button there is not need of any wrapper like below: (Ref: https://mui.com/material-ui/react-tooltip/)

![image](https://github.com/ChainSafe/Multix/assets/61174028/1d0c8d7f-971b-44b3-8db5-041401d47e0b)

About using onClick() directly into Identicon we can do that. Though the method will change to onCopy() but works fine.

![image](https://github.com/ChainSafe/Multix/assets/61174028/27585263-6511-4176-b83c-c5ca2da2e6f7)

Let me know if we should change that. Thanks!!

Tbaut replied:
Yes, let's use `onCopy` then. No need to wrap it.

dj-pundir replied:
Sorry but we have to use `<IconButton>` wrapper even if we use onCopy() handler. Because `Identicon` is an icon type not a button and `<Tooltip>` understands button only. I have committed the changes please check. Thanks!!

Tbaut replied:
The current state of the PR is not acceptable, it breaks the UI
![image](https://github.com/ChainSafe/Multix/assets/33178835/e33465ae-ab8e-4f34-aed9-f45fe972e4d3)

Tbaut replied:
You can change this IconButton by a div. Also please make sure the tooltip is on 1 line.
![image](https://github.com/ChainSafe/Multix/assets/33178835/bb2ebd77-49dc-4c2a-95c0-63a84f7843ea)

dj-pundir replied:
Please check now its using same style as before. Thanks!!

[screen-recorder-mon-may-22-2023-16-34-34.webm](https://github.com/ChainSafe/Multix/assets/61174028/80aa466c-29a5-4083-976c-0f8bc88ef5bc)

Tbaut commented on packages/ui/src/components/MultixIdenticon.tsx:
```diff
@@ -0,0 +1,45 @@
+import { IconTheme } from '@polkadot/react-identicon/types'
+
+const DEFAULT_PLACEMENT = 'top'
+const DEFAULT_TITLE = 'Copied'
```
Maybe it would be clearer to our users what is actually copied
```suggestion
const DEFAULT_TITLE = 'Address copied!'
```

dj-pundir replied:
Hey @Tbaut,  Is it always ADDRESS what we copy throughout the app?

dj-pundir replied:
Hey @Tbaut  Also Is there any way i can click on CREATE button here. I am sure we are using rococo testing environment here to send ROC etc.

![image](https://github.com/ChainSafe/Multix/assets/61174028/7e3b3f8e-d019-436a-8f90-c4bca14f2278)

Tbaut replied:
The error says you don't have enough ROC on the Signing account. Get some ROC using the faucet: https://matrix.to/#/#rococo-faucet:matrix.org

> Hey @Tbaut, Is it always ADDRESS what we copy throughout the app?

The copy action is from the `Identicon` component that we take from a library, so we don't handle the copy per se, and yet it'll always be the address copied in this case

dj-pundir replied:
> The error says you don't have enough ROC on the Signing account. Get some ROC using the faucet: https://matrix.to/#/#rococo-faucet:matrix.org


@Tbaut  it is showing choose app to continue. Can i choose any one of them?

![image](https://github.com/ChainSafe/Multix/assets/61174028/0c1a0678-f70a-480e-9573-af2cb48f702c)

asnaith replied:
I can answer this one @dj-pundir. Yep, you should be able to use any of them. Element seems to be the most popular.

Tbaut reviewed (changes requested):
Thank you so much for your contribution. This looks very good. I left some small comments, but nothing critical.

Tbaut commented:
Please make sure it's passing the lint and building without warning.

Tbaut reviewed (changes requested):
Please sure to run `yarn lint --fix` and resolve any issue.

dj-pundir commented:
> Please sure to run `yarn lint --fix` and resolve any issue.

Actually when i ran yarn lint --fix its showing so many problems at my end so i fixed the problems only for my files which i had changed. Not sure why this is happening. Thanks!!

![image](https://github.com/ChainSafe/Multix/assets/61174028/eeeef926-2fdb-4e95-97cd-e676e969a8bb)

tarunsamanta2k20 commented:
@Tbaut is this issue resolved?

Tbaut commented:
Please run `yarn formatAll` at the root of the package, and `yarn lint --fix` too. Thanks

Tbaut reviewed (approved):
Awesome, thanks a lot
